### PR TITLE
APIS-4687 - Change gatekeeper to save blank subscription fields values

### PR DESCRIPTION
**Pull request by adampridmore**
When a new subscription is made with no values.

### CODE REVIEW

File: app/connectors/SubscriptionFieldsConnector.scala [hunk 1]
```diff
@@ -60,16 +81,23 @@ abstract class AbstractSubscriptionFieldsConnector(implicit ec: ExecutionContext
...
-    } yield joinFieldValuesToDefinitions(definitions, fieldValues)
+    }  yield joinFieldValuesToDefinitions(definitions, fieldValues)
+  }
+
+  def fetchFieldDefinitions(apiContext: String, apiVersion: String)(implicit hc: HeaderCarrier): Future[Seq[SubscriptionFieldDefinition]] = {
+    val url = urlSubscriptionFieldDefinition(apiContext, apiVersion)
+    Logger.debug(s"fetchFieldDefinitions() - About to call $url in ${environment.toString}")
+    retry {
+      http.GET[ApiFieldDefinitions](url).map(response => response.fieldDefinitions.map(toDomain))
+    } recover recovery(Seq.empty[SubscriptionFieldDefinition])
```
Comment (AndySpaven) [inline app/connectors/SubscriptionFieldsConnector.scala]: Seems a bit of an excessive catch all.  At the very least only NonFatal should be caught.

File: app/services/ApplicationService.scala [hunk 2]
```diff
@@ -125,7 +125,7 @@ class ApplicationService @Inject()(sandboxApplicationConnector: SandboxApplicati
                                   version: VersionSubscription): Future[VersionSubscription] = {
 
       if (withFields) {
-        val apiContextVersion = ApiContextVersion(subscription.context, version.version.version)
+        val apiContextVersion = APIIdentifier(subscription.context, version.version.version)
```
Comment (AndySpaven) [inline app/services/ApplicationService.scala]: rename variable

File: app/services/ApplicationService.scala [hunk 3]
```diff
@@ -220,7 +220,46 @@ class ApplicationService @Inject()(sandboxApplicationConnector: SandboxApplicati
...
+      fieldDefinitions
+        .map(d => (d.name, ""))
+        .toMap
+    }
+
+    def ensureEmptyValuesWhenNoneExists(fieldDefinitions: Seq[SubscriptionFieldDefinition]): Future[HasSucceeded] = {
+      subscriptionFieldsService.fetchFieldsValues(application, fieldDefinitions, apiIdentifier)
+        .flatMap(values => {
+          if (!values.exists(field => field.value != "")) {
+            val x = subscriptionFieldsService.saveFieldValues(application, context, version, createEmptyFieldValues(fieldDefinitions))
```
Comment (AndySpaven) [inline app/services/ApplicationService.scala]: Remove this temporary variable [x] (I know it was temporary - I introduced it :-( )

File: app/connectors/SubscriptionFieldsConnector.scala (hunk 1 above)
Comment (peteslater-ee) [reply app/connectors/SubscriptionFieldsConnector.scala]: Moar explain pls

Comment (AndySpaven) [reply app/connectors/SubscriptionFieldsConnector.scala]: Ignore me - turns out the recovery function is the partial function and not the one I was expecting....  So it's fine.

Comment (peteslater-ee) [reply app/connectors/SubscriptionFieldsConnector.scala]: I see - glad I asked

**Pull request merged**